classes/page_group.php

```diff
@@ -0,0 +1,114 @@
+<?php
+// This file is part of Moodle - https://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+namespace tool_excimer;
+
+use core\persistent;
+
+/**
+ * Metadata for profile groups.
+ *
+ * @package   tool_excimer
+ * @author    Jason den Dulk <jasondendulk@catalyst-au.net>
+ * @copyright 2022, Catalyst IT
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class page_group extends persistent {
+
+    /** The name of the database table. */
+    public const TABLE = 'tool_excimer_page_groups';
+
+    /**
+     * Returns the fuzzy counts for duration.
+     *
+     * @return array
+     */
+    protected function get_fuzzydurationcounts(): array {
+        $raw = $this->raw_get('fuzzydurationcounts');
+        if (empty($raw)) {
+            return [];
+        }
+
+        return json_decode($raw, true);
+    }
+
+    /**
+     * Sets the fuzzy duration counts.
+     *
+     * @param array $data
+     */
+    protected function set_fuzzydurationcounts(array $data) {
+        $this->raw_set('fuzzydurationcounts', json_encode($data));
+    }
+
+    /**
+     * Records fuzzy count metadata about a page group.
+     *
+     * @param profile $profile The profile to pull the information from.
+     */
+    public static function record_fuzzy_counts(profile $profile) {
+
+        // Get the profile group record, creating a new one if one does not yet exist.
+        $month = userdate(time() - 360, '%Y%m'); // YYYYMM format.
+        $profilegroup = self::get_record(['name' => $profile->get('groupby'), 'month' => $month]);
```

Comment:
Ultimately I want this also in MUC so we don't hit the db at all but save implementing this until right at the end when we know its all working correctly